server/models/post.model.js

```diff
@@ -152,56 +163,67 @@ PostSchema.statics = {
     if (type === 'top') {
       sort = { score: -1 };
     }
-
-    // return this.find().count()
-    //   .then((count) => {
-    //     numberOfPages = Math.floor(count / limit);
-    //
-    //     return this.find(query)
-    //       .sort(sort)
-    //       .limit(limit)
-    //       .exec();
-    //   })
-
-    return this.find(query, 'content title date mp3 link score featuredImage upvoted downvoted tags categories')
+    const queryPromise = this.find(query, 'content title date mp3 link score featuredImage upvoted downvoted tags categories')
       .sort(sort)
-      .limit(limitOption)
-      .then((postsFound) => {
+      .limit(limitOption);
+
+    if (!user) {
+      return queryPromise.then((postsFound) => {
         posts = postsFound;
         // Flip direct back
         if (dateDirection === 1) {
           posts.reverse();
         }
-        if (!user) return posts;
+        return postsFound;
+      });
+    }
 
+    return queryPromise
+      .populate({
+        path: 'favoritedByUser',
+        select: 'active',
+        match: { userId: user._id }
+      }).exec()
+      .then((postsFound) => {
+        posts = postsFound;
+        // Flip direct back
+        if (dateDirection === 1) {
+          posts.reverse();
+        }
         const postIds = posts.map((post) => { //eslint-disable-line
           return post._id;
         });
 
-        return Vote.find({$or: [
-          {
-            userId: user._id,
-            postId:  {$in: postIds},
-          },            {
-            userId: user._id,
-            entityId: {$in: postIds},
-          },
-        ]}).exec();
+        return Vote.find({
+          $or: [
+            {
+              userId: user._id,
+              postId: { $in: postIds },
+            }, {
+              userId: user._id,
+              entityId: { $in: postIds },
+            },
+          ]
+        }).exec();
       })
       .then((votes) => {
-        if (!user) return posts;
         const voteMap = {};
         for (let index in votes) { // eslint-disable-line
           const vote = votes[index];
-          const voteKey = vote.postId?  vote.postId : vote.entityId;
+          const voteKey = vote.postId ? vote.postId : vote.entityId;
           voteMap[voteKey] = vote;
         }
 
         const updatedPosts = [];
         for (let index in posts) { // eslint-disable-line
           const post = posts[index].toObject();
+          post.bookmarked = false;
           post.upvoted = false;
           post.downvoted = false;
+          // virtual is lost in toObject()
+          // even if the following set in schema: { toJSON: { virtuals: true } });
+          const { favoritedByUser } = posts[index];
```

Comment:
Ahh, I wonder if using .lean() on the original query would do the same :/

Reply:
Guess not but maybe this: https://github.com/vkarpov15/mongoose-lean-virtuals

Reply:
Something worth exploring. Definitely recognize the code is not perfect. But if we adopt virtuals more fully I figure we'll refactor later to incorporate vote.